File: src/hir.rs
//! High level IR

#![allow(dead_code)]

use crate::ast::{Attr, BinOp, CastKind, IntKind, Linkage, Param, Span, Ty, UnOp};

#[derive(Clone, Debug)]
pub enum TExpr {
    Int {
        val: u128,
        ty: Ty,
        span: Span,
    },
    IntLit {
        val: i128,
        span: Span,
    },
    Bool {
        val: bool,
        span: Span,
    },
    NullPtr {
        ty: Ty,
        span: Span,
    },
    NullLit {
        span: Span,
    },
    StrLit {
        index: usize,
        span: Span,
    },
    Ident {
        name: String,
        ty: Ty,
        span: Span,
    },
    Unary {
        op: UnOp,
        e: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    Binary {
        op: BinOp,
        l: Box<TExpr>,
        r: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    Call {
        target: CallTarget,
        args: Vec<TExpr>,
        ret: Ty,
        span: Span,
    },
    Cast {
        kind: CastKind,
        e: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    Index {
        base: Box<TExpr>,
        idx: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    Field {
        base: Box<TExpr>,
        name: String,
        ty: Ty,
        offset: u64,
        span: Span,
    },
    Deref {
        e: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    AddrOf {
        e: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    PtrAdd {
        bytes: bool,
        ptr: Box<TExpr>,
        off: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    SizeOf {
        ty: Ty,
        size: u64,
        span: Span,
    },
    AlignOf {
        ty: Ty,
        align: u64,
        span: Span,
    },
    OffsetOf {
        ty: Ty,
        field: String,
        offset: u64,
        span: Span,
    },
    Splat {
        e: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    Ternary {
        cond: Box<TExpr>,
        then: Box<TExpr>,
        els: Box<TExpr>,
        ty: Ty,
        span: Span,
    },
    ConstArr {
        name: String,
        ty: Ty,
        span: Span,
    },
    StructLit {
        name: String,
        elems: Vec<TExpr>,
        ty: Ty,
        span: Span,
    },
}

#[derive(Clone, Debug)]
pub enum CallTarget {
    User(String),
    Builtin(Builtin),
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Builtin {
    Shl,
    Lshr,
    Ashr,
    Rotl,
    Rotr,
    Popcount,
    Ctz,
    Clz,
}

#[derive(Clone, Debug)]
pub enum TStmt {
    Let {
        name: String,
        ty: Ty,
        init: Option<Box<TExpr>>,
        span: Span,
    },
    Assign {
        target: TAssignTarget,
        value: Box<TExpr>,
        span: Span,
    },
    Return(Option<Box<TExpr>>, Span),
    If {
        cond: Box<TExpr>,
        then: Vec<TStmt>,
        els: Option<Vec<TStmt>>,
        span: Span,
    },
    While {
        cond: Box<TExpr>,
        body: Vec<TStmt>,
        span: Span,
    },
    Break(Span),
    Continue(Span),
    Defer(Vec<TStmt>, Span),
    Expr(Box<TExpr>, Span),
    Block(Vec<TStmt>),
    For {
        init: Option<Box<TStmt>>,
        cond: Option<Box<TExpr>>,
        step: Option<Box<TStmt>>,
        body: Vec<TStmt>,
        span: Span,
    },
}

#[derive(Clone, Debug)]
pub enum TAssignTarget {
    Ident(String),
    Deref(Box<TExpr>),
    Index { base: Box<TExpr>, idx: Box<TExpr> },
    Field { base: Box<TExpr>, name: String, offset: u64 },
}

pub type StructDecl = (String, Vec<(String, Ty)>, Option<u32>);

pub fn layout(structs: &[StructDef], ty: &Ty) -> Result<(u64, u64), String> {
    match ty {
        Ty::Bool => Ok((1, 1)),
        Ty::Int(IntKind::U8 | IntKind::I8) => Ok((1, 1)),
        Ty::Int(IntKind::U16 | IntKind::I16) => Ok((2, 2)),
        Ty::Int(IntKind::U32 | IntKind::I32) => Ok((4, 4)),
        Ty::Int(
            IntKind::U64 | IntKind::I64 | IntKind::Usize | IntKind::Isize,
        ) => Ok((8, 8)),
        Ty::Ptr(_) => Ok((8, 8)),
        Ty::Array(n, elem) => {
            let (es, ea) = layout(structs, elem)?;
            let sz = n
                .checked_mul(es)
                .ok_or_else(|| format!("array size {n} x {es} overflows"))?;
            Ok((sz, ea))
        }
        Ty::Vec(n, elem) => {
            let (es, ea) = layout(structs, elem)?;
            let sz = n
                .checked_mul(es)
                .ok_or_else(|| format!("vector size {n} x {es} overflows"))?;
            Ok((sz, ea))
        }
        Ty::Struct(name) => {
            let s = structs
                .iter()
                .find(|s| &s.name == name)
                .ok_or_else(|| format!("unknown struct `{name}`"))?;
            Ok((s.size, s.align))
        }
        Ty::Opaque(name) => Err(format!("opaque type `{name}` has no known layout")),
        Ty::Void => Err("void has no object layout".into()),
    }
}

pub fn compute_struct_layout(
    name: &str,
    all_structs: &[StructDecl],
) -> Result<(Vec<FieldDef>, u64, u64), String> {
    fn resolve(
        name: &str,
        all: &[StructDecl],
        memo: &mut std::collections::HashMap<String, (u64, u64)>,
        visiting: &mut std::collections::HashSet<String>,
    ) -> Result<(u64, u64), String> {
        if let Some(v) = memo.get(name) {
            return Ok(*v);
        }
        if !visiting.insert(name.to_string()) {
            return Err(format!(
                "recursive struct layout involving `{name}` (by-value self-reference)"
            ));
        }
        let result = (|| {
            let (_, fields, align) = all
                .iter()
                .find(|(n, _, _)| n == name)
                .ok_or_else(|| format!("unknown struct `{name}`"))?;
            let mut offset = 0u64;
            let mut max_align = 1u64;
            for (_, fty) in fields {
                let (fsize, falign) = type_layout(fty, all, memo, visiting)?;
                max_align = max_align.max(falign);
                offset = round_up(offset, falign).checked_add(fsize)
                    .ok_or_else(|| format!("struct `{name}` size overflows"))?;
            }
            let align = align.map(|a| a as u64).unwrap_or(max_align).max(max_align);
            let size = round_up(offset, align);
            Ok((size, align))
        })();
        visiting.remove(name);
        if let Ok((size, align)) = result {
            memo.insert(name.to_string(), (size, align));
        }
        result
    }

    fn type_layout(
        ty: &Ty,
        all: &[StructDecl],
        memo: &mut std::collections::HashMap<String, (u64, u64)>,
        visiting: &mut std::collections::HashSet<String>,
    ) -> Result<(u64, u64), String> {
        match ty {
            Ty::Bool => Ok((1, 1)),
            Ty::Int(k) => Ok(match k {
                IntKind::U8 | IntKind::I8 => (1, 1),
                IntKind::U16 | IntKind::I16 => (2, 2),
                IntKind::U32 | IntKind::I32 => (4, 4),
                IntKind::U64 | IntKind::I64 | IntKind::Usize | IntKind::Isize => (8, 8),
            }),
            Ty::Ptr(_) => Ok((8, 8)),
            Ty::Array(n, e) => {
                let (es, ea) = type_layout(e, all, memo, visiting)?;
                Ok((n.checked_mul(es).ok_or_else(|| format!("array size {n} x {es} overflows"))?, ea))
            }
            Ty::Vec(n, e) => {
                let (es, ea) = type_layout(e, all, memo, visiting)?;
                Ok((n.checked_mul(es).ok_or_else(|| format!("vector size {n} x {es} overflows"))?, ea))
            }
            Ty::Struct(s) => resolve(s, all, memo, visiting),
            Ty::Opaque(name) => Err(format!("opaque type `{name}` cannot be used by value")),
            Ty::Void => Err("void cannot be used as stored object type".into()),
        }
    }

    let mut memo = std::collections::HashMap::new();
    let mut visiting = std::collections::HashSet::new();
    resolve(name, all_structs, &mut memo, &mut visiting)?;

    let (_, fields, _) = all_structs
        .iter()
        .find(|(n, _, _)| n == name)
        .expect("resolved");
    let mut fields_out = Vec::new();
    let mut offset = 0u64;
    for (fname, fty) in fields {
        let (fsize, falign) = type_layout(fty, all_structs, &mut memo, &mut visiting)?;
        offset = round_up(offset, falign);
        fields_out.push(FieldDef {
            name: fname.clone(),
            ty: fty.clone(),
            offset,
        });
        offset = offset.checked_add(fsize).ok_or_else(|| format!("struct `{name}` size overflows"))?;
    }
    let (size, align) = *memo.get(name).expect("resolved");
    Ok((fields_out, size, align))
}

fn round_up(x: u64, align: u64) -> u64 {
    if align == 0 {
        return x;
    }
    (x + align - 1) & !(align - 1)
}

pub fn is_addressable(&self) -> bool {
    match self {
        TExpr::Ident { .. } => true,
        TExpr::ConstArr { .. } => true,
        TExpr::Deref { .. } => true,
        TExpr::Index { base, .. } => match base.ty() {
            Ty::Ptr(..) => true,
            Ty::Array(..) => base.is_addressable(),
            Ty::Vec(..) => false,
            _ => false,
        },
        TExpr::Field { base, .. } => base.is_addressable(),
        _ => false,
    }
}

pub fn is_assignable(&self) -> bool {
    match self {
        TExpr::Ident { .. } => true,
        TExpr::ConstArr { .. } => true,
        TExpr::Deref { .. } => true,
        TExpr::Index { base, .. } => match base.ty() {
            Ty::Ptr(..) => true,
            Ty::Array(..) | Ty::Vec(..) => base.is_addressable(),
            _ => false,
        },
        TExpr::Field { base, .. } => base.is_addressable(),
        _ => false,
    }
}

#[derive(Clone, Debug)]
pub struct Program {
    pub structs: Vec<StructDef>,
    pub consts: Vec<ConstDef>,
    pub const_arrs: Vec<ConstArrDef>,
    pub funcs: Vec<FuncDef>,
    pub strings: Vec<String>,
}

#[derive(Clone, Debug)]
pub struct StructDef {
    pub name: String,
    pub fields: Vec<FieldDef>,
    pub size: u64,
    pub align: u64,
}

#[derive(Clone, Debug)]
pub struct FieldDef {
    pub name: String,
    pub ty: Ty,
    pub offset: u64,
}

#[derive(Clone, Debug)]
pub struct ConstDef {
    pub name: String,
    pub ty: Ty,
    pub bits: u128,
}

#[derive(Clone, Debug)]
pub struct ConstArrDef {
    pub name: String,
    pub ty: Ty,
    pub elems: Vec<u128>,
}

#[derive(Clone, Debug)]
pub struct FuncDef {
    pub name: String,
    pub params: Vec<Param>,
    pub ret: Ty,
    pub body: Vec<TStmt>,
    pub attrs: Vec<Attr>,
    pub linkage: Linkage,
    pub variadic: bool,
    pub asm_body: Option<String>,
}

File: src/hir.ml
type int_kind = U8 | U16 | U32 | I8 | I16 | I32 | I64 | U64 | Usize | Isize

type ty =
  | Bool
  | Int of int_kind
  | Ptr of ty
  | ConstPtr of ty
  | Array of int * ty
  | Vec of int * ty
  | Struct of string
  | Opaque of string
  | Void

type field = { name : string; ty : ty; offset : int }
type struct_def = { name : string; fields : field list; size : int; align : int }
type const_def = { name : string; ty : ty; bits : int64 }
type const_arr_def = { name : string; ty : ty; elems : int64 list }
type func_sig = { params : (string * ty) list; ret : ty; variadic : bool }
type local = { name : string; ty : ty; id : int }
type linkage = Internal | External_c
type builtin = Rotl | Rotr | Popcount | Ctz | Clz | Add_sat | Sub_sat
type call_target = User of string | Builtin of builtin

type expr =
  | EInt of int64 * ty * Span.t
  | EBool of bool * Span.t
  | EVector of int64 list * ty * Span.t
  | Null of ty * Span.t
  | EString of int * Span.t
  | Local of local * Span.t
  | Unary of Ast.unop * expr * ty * Span.t
  | Binary of Ast.binop * expr * expr * ty * Span.t
  | Call of call_target * expr list * ty * Span.t
  | Cast of Ast.cast_kind * expr * ty * Span.t
  | Index of expr * expr * ty * Span.t
  | Field of expr * string * ty * int * Span.t
  | Deref of expr * ty * Span.t
  | Address of expr * ty * Span.t
  | Ptr_add of bool * expr * expr * ty * Span.t
  | Sizeof of ty * int * Span.t
  | Alignof of ty * int * Span.t
  | Offsetof of ty * string * int * Span.t
  | Splat of expr * ty * Span.t
  | Ternary of expr * expr * expr * ty * Span.t
  | Const_array of string * ty * Span.t
  | Struct_lit of string * expr list * ty * Span.t

type assign_target =
  | ALocal of local
  | ADeref of expr
  | AIndex of expr * expr
  | AField of expr * string * int

type stmt =
  | Let of local * expr option * Span.t
  | Assign of assign_target * expr * Span.t
  | Compound_assign of assign_target * Ast.binop * expr * ty * Span.t
  | Return of expr option * Span.t
  | If of expr * stmt list * stmt list option * Span.t
  | While of expr * stmt list * Span.t
  | For of stmt option * expr option * stmt option * stmt list * Span.t
  | Switch of expr * (expr * stmt list) list * stmt list option * Span.t
  | Break of Span.t
  | Continue of Span.t
  | Defer of stmt list * Span.t
  | Expr of expr * Span.t
  | Block of stmt list * Span.t

type func_body = Declaration | Statements of stmt list | Asm of string

type func = {
  name : string;
  params : local list;
  ret : ty;
  body : func_body;
  linkage : linkage;
  variadic : bool;
}

type program = {
  structs : struct_def list;
  consts : const_def list;
  const_arrays : const_arr_def list;
  funcs : func list;
  strings : string list;
}

let rec ty_equal a b =
  match (a, b) with
  | Bool, Bool | Void, Void -> true
  | Int a, Int b -> a = b
  | Ptr a, Ptr b -> ty_equal a b
  | ConstPtr a, ConstPtr b -> ty_equal a b
  | Array (na, a), Array (nb, b) | Vec (na, a), Vec (nb, b) -> na = nb && ty_equal a b
  | Struct a, Struct b | Opaque a, Opaque b -> a = b
  | _ -> false

let rec ty_name = function
  | Bool -> "bool"
  | Void -> "void"
  | Int U8 -> "u8"
  | Int U16 -> "u16"
  | Int U32 -> "u32"
  | Int U64 -> "u64"
  | Int I8 -> "i8"
  | Int I16 -> "i16"
  | Int I32 -> "i32"
  | Int I64 -> "i64"
  | Int Usize -> "usize"
  | Int Isize -> "isize"
  | Ptr t -> "ptr[" ^ ty_name t ^ "]"
  | ConstPtr t -> "ptr[const " ^ ty_name t ^ "]"
  | Array (n, t) -> Printf.sprintf "arr[%d, %s]" n (ty_name t)
  | Vec (n, t) -> Printf.sprintf "vec[%d, %s]" n (ty_name t)
  | Struct n | Opaque n -> n

let expr_ty = function
  | EInt (_, t, _)
  | EVector (_, t, _)
  | Unary (_, _, t, _)
  | Binary (_, _, _, t, _)
  | Call (_, _, t, _)
  | Cast (_, _, t, _)
  | Index (_, _, t, _)
  | Field (_, _, t, _, _)
  | Deref (_, t, _)
  | Address (_, t, _)
  | Ptr_add (_, _, _, t, _)
  | Splat (_, t, _)
  | Ternary (_, _, _, t, _)
  | Const_array (_, t, _)
  | Struct_lit (_, _, t, _) ->
      t
  | Local (local, _) -> local.ty
  | EBool _ -> Bool
  | Null (t, _) -> t
  | EString _ -> ConstPtr (Int U8)
  | Sizeof _ | Alignof _ | Offsetof _ -> Int Usize

let expr_span = function
  | EInt (_, _, s)
  | EVector (_, _, s)
  | EBool (_, s)
  | Null (_, s)
  | EString (_, s)
  | Local (_, s)
  | Unary (_, _, _, s)
  | Binary (_, _, _, _, s)
  | Call (_, _, _, s)
  | Cast (_, _, _, s)
  | Index (_, _, _, s)
  | Field (_, _, _, _, s)
  | Deref (_, _, s)
  | Address (_, _, s)
  | Ptr_add (_, _, _, _, s)
  | Sizeof (_, _, s)
  | Alignof (_, _, s)
  | Offsetof (_, _, _, s)
  | Splat (_, _, s)
  | Ternary (_, _, _, _, s)
  | Const_array (_, _, s)
  | Struct_lit (_, _, _, s) ->
      s

type flow_summary = {
  falls_through : bool;
  returns : bool;
  breaks : bool;
  continues : bool;
}

let flowing =
  { falls_through = true; returns = false; breaks = false; continues = false }

let choose_flow left right =
  {
    falls_through = left.falls_through || right.falls_through;
    returns = left.returns || right.returns;
    breaks = left.breaks || right.breaks;
    continues = left.continues || right.continues;
  }

let sequence_flow left right =
  {
    falls_through = left.falls_through && right.falls_through;
    returns = left.returns || (left.falls_through && right.returns);
    breaks = left.breaks || (left.falls_through && right.breaks);
    continues = left.continues || (left.falls_through && right.continues);
  }

let cleanup_flow cleanup exits =
  {
    falls_through = cleanup.falls_through && exits.falls_through;
    returns = cleanup.falls_through && exits.returns;
    breaks = cleanup.falls_through && exits.breaks;
    continues = cleanup.falls_through && exits.continues;
  }

let condition_is_true = function EBool (true, _) -> true | _ -> false

let rec stmt_flow = function
  | Return _ -> { flowing with falls_through = false; returns = true }
  | Break _ -> { flowing with falls_through = false; breaks = true }
  | Continue _ -> { flowing with falls_through = false; continues = true }
  | Block (body, _) -> block_flow body
  | If (_, then_body, else_body, _) ->
      choose_flow (block_flow then_body)
        (match else_body with None -> flowing | Some body -> block_flow body)
  | Switch (_, arms, default, _) ->
      let branches = List.map (fun (_, body) -> block_flow body) arms in
      let branches =
        match default with
        | None -> flowing :: branches
        | Some body -> block_flow body :: branches
      in
      List.fold_left choose_flow { flowing with falls_through = false } branches
  | While (condition, body, _) ->
      loop_flow (condition_is_true condition) (block_flow body)
  | For (init, condition, step, body, _) ->
      let prefix =
        match init with None -> flowing | Some statement -> stmt_flow statement
      in
      let iteration =
        sequence_flow (block_flow body)
          (match step with None -> flowing | Some statement -> stmt_flow statement)
      in
      let unconditional =
        match condition with
        | None -> true
        | Some expression -> condition_is_true expression
      in
      sequence_flow prefix (loop_flow unconditional iteration)
  | Defer (body, _) -> cleanup_flow (block_flow body) flowing
  | Let _ | Assign _ | Compound_assign _ | Expr _ -> flowing

and block_flow = function
  | [] -> flowing
  | Defer (body, _) :: rest -> cleanup_flow (block_flow body) (block_flow rest)
  | statement :: rest -> sequence_flow (stmt_flow statement) (block_flow rest)

and loop_flow unconditional body =
  {
    falls_through = (not unconditional) || body.breaks;
    returns = body.returns;
    breaks = false;
    continues = false;
  }

let ( let* ) r f = match r with Error e -> Error e | Ok x -> f x

let add_size left right =
  if left < 0 || right < 0 || left > max_int - right then
    Error "aggregate size overflows"
  else Ok (left + right)

let int_bytes ?(target = Target_layout.current) = function
  | U8 | I8 -> 1
  | U16 | I16 -> 2
  | U32 | I32 -> 4
  | U64 | I64 -> 8
  | Usize | Isize -> target.Target_layout.pointer_size

let int_layout target k = Target_layout.integer target (int_bytes ~target k * 8)

let scalar_bits target = function
  | Bool -> Ok 1
  | Int k -> Ok (int_bytes ~target k * 8)
  | Ptr _ | ConstPtr _ -> Ok (target.Target_layout.pointer_size * 8)
  | _ -> Error "vector element type must be a scalar"

let layout ?(target = Target_layout.current) structs ty =
  let rec go visiting = function
    | Bool -> Target_layout.integer target 1
    | Int k -> int_layout target k
    | Ptr _ | ConstPtr _ -> Target_layout.pointer target
    | Array (n, t) ->
        let* s, a = go visiting t in
        let* size = Target_layout.multiply_size n s in
        Ok (size, a)
    | Vec (n, t) ->
        let* bits = scalar_bits target t in
        Target_layout.vector target n bits
    | Void -> Error "void has no object layout"
    | Opaque n -> Error ("opaque type `" ^ n ^ "` has no layout")
    | Struct n -> (
        if List.mem n visiting then
          Error (Printf.sprintf "recursive by-value struct `%s`" n)
        else
          match List.find_opt (fun (s : struct_def) -> s.name = n) structs with
          | None -> Error (Printf.sprintf "unknown struct `%s`" n)
          | Some (s : struct_def) -> Ok (s.size, s.align))
  in
  go [] ty

type struct_layout_cache = {
  target : Target_layout.t;
  decls : (string * (string * ty) list * int option) list;
  definitions : (string, struct_def) Hashtbl.t;
}

let struct_layout_cache ?(target = Target_layout.current) decls =
  { target; decls; definitions = Hashtbl.create (List.length decls) }

let compute_struct_cached cache name =
  let target = cache.target in
  let decls = cache.decls in
  let rec calc visiting n =
    match Hashtbl.find_opt cache.definitions n with
    | Some definition -> Ok (definition.fields, definition.size, definition.align)
    | None -> (
        if List.mem n visiting then
          Error (Printf.sprintf "recursive by-value struct `%s`" n)
        else
          match List.find_opt (fun (x, _, _) -> x = n) decls with
          | None -> Error (Printf.sprintf "unknown struct `%s`" n)
          | Some (_, fields, explicit) ->
              let rec each off maxa out = function
                | [] ->
                    let align = max maxa (Option.value ~default:1 explicit) in
                    let* size = Target_layout.round_up_size off align in
                    let definition = { name = n; fields = List.rev out; size; align } in
                    Hashtbl.replace cache.definitions n definition;
                    Ok (definition.fields, definition.size, definition.align)
                | (fname, fty) :: rest ->
                    let* size, align =
                      match fty with
                      | Struct sn ->
                          let* _, sz, al = calc (n :: visiting) sn in
                          Ok (sz, al)
                      | _ -> field_layout (n :: visiting) fty
                    in
                    let* next = Target_layout.round_up_size off align in
                    let* next_offset = add_size next size in
                    each next_offset (max maxa align)
                      ({ name = fname; ty = fty; offset = next } :: out)
                      rest
              in
              each 0 1 [] fields)
  and field_layout visiting = function
    | Bool -> Target_layout.integer target 1
    | Int k -> int_layout target k
    | Ptr _ | ConstPtr _ -> Target_layout.pointer target
    | Void -> Error "void has no object layout"
    | Opaque n -> Error (Printf.sprintf "opaque type `%s` has no layout" n)
    | Struct n ->
        let* _, s, a = calc visiting n in
        Ok (s, a)
    | Array (n, t) ->
        let* s, a = field_layout visiting t in
        let* size = Target_layout.multiply_size n s in
        Ok (size, a)
    | Vec (n, t) ->
        let* bits = scalar_bits target t in
        Target_layout.vector target n bits
  in
  let* fields, size, align = calc [] name in
  Ok { name; fields; size; align }

let compute_struct ?target decls name =
  compute_struct_cached (struct_layout_cache ?target decls) name

let render p =
  let one_struct (s : struct_def) =
    Printf.sprintf "struct %s size=%d align=%d" s.name s.size s.align
  in
  let one_fn f =
    Printf.sprintf "fn %s(%s) %s" f.name
      (String.concat ", "
         (List.map
            (fun (local : local) -> local.name ^ ":" ^ ty_name local.ty)
            f.params))
      (ty_name f.ret)
  in
  String.concat "\n" (List.map one_struct p.structs @ List.map one_fn p.funcs)
